openmeter/billing/invoiceline.go

```diff
@@ -101,28 +106,220 @@ func (i AssociateLinesToInvoiceAdapterInput) Validate() error {
 
 type UpdateInvoiceLineAdapterInput billingentity.Line
 
-type GetInvoiceLineInput struct {
-	Namespace string
-	LineID    string
-	InvoiceID string
+type UpdateInvoiceLineInput struct {
+	Invoice   billingentity.InvoiceID
+	UpdatedAt time.Time
+
+	LineBase   UpdateInvoiceLineBaseInput
+	UsageBased UpdateInvoiceLineUsageBasedInput
+	FlatFee    UpdateInvoiceLineFlatFeeInput
 }
 
-func (g GetInvoiceLineInput) Validate() error {
-	if g.Namespace == "" {
-		return errors.New("namespace is required")
+func (u UpdateInvoiceLineInput) Validate() error {
+	var outErr error
+	if err := u.LineBase.Validate(); err != nil {
+		outErr = errors.Join(outErr, err)
 	}
 
-	if g.LineID == "" {
-		return errors.New("line id is required")
+	if u.Invoice.ID == "" {
+		outErr = errors.Join(outErr, billingentity.ValidationWithFieldPrefix("id", billingentity.ErrFieldRequired))
 	}
 
-	if g.InvoiceID == "" {
-		return errors.New("invoice id is required")
+	if u.Invoice.Namespace == "" {
+		outErr = errors.Join(outErr, billingentity.ValidationWithFieldPrefix("namespace", billingentity.ErrFieldRequired))
+	}
+
+	if u.UpdatedAt.IsZero() {
```

Comment:
isn't updated at set at create?

Reply:
This is for conflict detection. I will remove this conflict detection in the next PR or make it optional backend logic only.

Right now each update mandates that you specify the current db version of UpdatedAt, so that we are not overwriting anything with stale data. This helps a lot in tests, but can be a bit daunting for API usage, so discussing with @tothandras I will drop this as part of the API refactor.